/*

Copyright(c) 2017 Cedric Jimenez

This file is part of Open-Vario Simulator.

Open-Vario Simulator is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Open-Vario Simulator is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with Open-Vario Simulator.  If not, see <http://www.gnu.org/licenses/>.

*/

syntax = "proto3";
option optimize_for = LITE_RUNTIME;

package open_vario;


/* Connection response message */
message ConnectResponse
{
    /* Indicates if the connection has been accepted */
    bool accept = 1;
};

/* Disconnection response message */
message DisconnectResponse
{};

/* List sensor response message */
message ListSensorsResponse
{
    /* Sensor types */
    enum SensorType 
    {
        /* Unknown sensor */
        ST_UNKNOWN = 0;
        /* Pressure sensor */
        ST_PRESSURE = 1;
        /* Temperature sensor */
        ST_TEMPERATURE = 2;
        /* Altitude sensor */
        ST_ALTITUDE = 4;
        /* GNSS sensor */
        ST_GNSS = 8;
    };

    /* Sensor value types */
    enum SensorValueType 
    {
        /* Unknown */
        SVT_UNKNOWN = 0;
        /* Unsigned integer */
        SVT_UINT = 1;
        /* Signed integer */
        SVT_INT = 2;
        /* Single precision floating point */
        SVT_FLOAT = 3;
        /* Double precision floating point */
        SVT_DOUBLE = 4;
        /* String */
        SVT_STRING = 5;
        /* Boolean */
        SVT_BOOL = 6;
    };

    /* Sensor description message */
    message Sensor
    {
        /* Sensor's id */
        uint32 id = 1;
        /* Sensor's name */
        string name = 2;
        /* Sensor's type */
        SensorType type = 3;
        /* Sensor's value type */
        SensorValueType value_type = 4;
    };
    /* List of available sensors */
    repeated Sensor sensors = 1;
};

/* Update sensor response message */
message UpdateSensorResponse
{
    /* Indicates if the sensor's value has been updated */
    bool success = 1; 
};

/* Ping response message */
message PingResponse
{
    /* Ping number */
    uint32 number = 1;
};

/* Configuration value write response message */
message ConfigValueWriteResponse
{
    /* Indicates if the configuration value has been written */
    bool success = 1;
};

/* Configuration value read response message */
message ConfigValueReadResponse
{
    /* Indicates if the configuration value has been read */
    bool success = 1;
    /* Value group name */
    string value_group_name = 2;
    /* Value name */
    string value_name = 3;
    /* Value type */
    string value_type = 4;
    /* Value size */
    uint32 value_size = 5;
    /* Indicate if the value has min and max values */
    bool has_min_max = 6;
    /* Indicate if the value will be taken into account at reset only */
    bool is_reset_only = 7;
    /* Configuration value */
    oneof Values
    {
        /* Value as unsigned integer */
        uint32 uint_value = 10;
        /* Value as integer */
        sint32 int_value = 11;
        /* Value as single precision floating point value */
        float float_value = 12;
        /* Value as double precision floating point value */
        double double_value = 13;
        /* Value as string */
        string string_value = 14;
        /* Value as boolean */
        bool bool_value = 15;
    };
    /* Configuration min value */
    oneof MinValues
    {
        /* Value as unsigned integer */
        uint32 uint_min_value = 20;
        /* Value as integer */
        sint32 int_min_value = 21;
        /* Value as single precision floating point value */
        float float_min_value = 22;
        /* Value as double precision floating point value */
        double double_min_value = 23;
        /* Value as string */
        string string_min_value = 24;
        /* Value as boolean */
        bool bool_min_value = 25;
    };
    /* Configuration max value */
    oneof MaxValues
    {
        /* Value as unsigned integer */
        uint32 uint_max_value = 30;
        /* Value as integer */
        sint32 int_max_value = 31;
        /* Value as single precision floating point value */
        float float_max_value = 32;
        /* Value as double precision floating point value */
        double double_max_value = 33;
        /* Value as string */
        string string_max_value = 34;
        /* Value as boolean */
        bool bool_max_value = 35;
    };
};

/* Simulator response message */
message SimuResponse
{ 
    /* Available responses */
    oneof Responses 
    {
        /* Connection response */
        ConnectResponse connect = 1;
        /* Disconnection response */
        DisconnectResponse disconnect = 2;
        /* List sensors response */
        ListSensorsResponse list_sensors = 3;
        /* Update sensor response */
        UpdateSensorResponse update_sensor = 4;
        /* Ping response */
        PingResponse ping = 5;
        /* Configuration value write response */
        ConfigValueWriteResponse config_write = 6;
        /* Configuration value read response */
        ConfigValueReadResponse config_read = 7;
    };
};
